### ISSUE TITLE
Search does not produce results, but map does

### ISSUE THREAD

**Issue by ceever** [osmandapp/OsmAnd#16097]
### Description
When searching for "dm" for instance to find drug stores of the DM chain in Germany, the search does not display any results. However, when clicking "Show on map" there are clearly many results in near distance.

See screenshots:
<p style="vertical-align:top;white-space:nowrap;">
<img src="https://user-images.githubusercontent.com/40247591/210009032-ea3e1277-4fdd-4faf-8bae-313c7d1636ea.jpg" width="300"/> &nbsp; <img src="https://user-images.githubusercontent.com/40247591/210009043-9ad78eaf-254c-4e79-9b84-427f9ec647b8.jpg" width="300"/>
</p>

### How to reproduce?
Go to the center of Berlin.
Zoom to include at least 5 km of surroundings.
Search for "dm" >> observe no results.
Click "Show on map" >> observe results.

### Your Environment
OsmAnd Version: 4.3.3
Android/iOS version: Android 9
Device model: Samsung S8

**Comment by scaidermern** [comment]
Searching for "dm " (with a space at the end) yields results but is not intuitive, of course.

**Comment by sonora** [comment]
Yes, I think we start searching only after 3 characters have been typed. Before that the algorithm would burn battery power to produce masses of 1 or 2 character hit results, which are mostly useless (your "dm" may be an exception... :wink: ).

**Comment by scaidermern** [comment]
> Yes, I think we start searching only after 3 characters have been typed. Before that the algorithm would burn battery power to produce masses of 1 or 2 character hit results, which are mostly useless

I always wondered why OsmAnd immediately starts searching while the user is still typing. *This* burns battery power and is even quite annoying. Just wait 0.5 seconds after the last input before staring to search. Then the 2 character limit can be dropped easily and *this* issue is resolved.

**Comment by ceever** [comment]
@scaidermern is right. This instant search does only work for quick devices anyhow. I already had cases where I still waited a few seconds before something is displayed, namely as in the following example.

Indeed, "dm" does come up with something when I search somewhere else, namely with personal favourists and tracks, see:
<p style="vertical-align:top;white-space:nowrap;">
<img src="https://user-images.githubusercontent.com/40247591/210106957-d1d231fa-5bd2-4aba-bc97-9fcec9b76fef.jpg" width="300"/>
</p>

This is a little inconsistent. I would suggest, we just drop the instant search for very short searches, even for one letter and display something after 1-2 seconds. 👍

Best

**Comment by vshcherb** [comment]
> Searching for "dm " (with a space at the end) yields results but is not intuitive, of course.

We don't do partial search with 2 letter, so user needs to hit Search button to add " ". We're discussing to remove partial search results if there is enough delay and do complete search but it's not clear whether it's better or not

**Comment by scaidermern** [comment]
Unfortunately current behaviour is not very intuitive. The user types two letters and OsmAnd starts to display search results. At this point the user expects the list of search results to be complete. How should the user know that the search button must be pressed just to improve the search results? If there are already results visible then the user never expects that the button needs to be pressed in order have different results.

Is there any other app that has the same behavior? 

How about a toggle button to switch between partial and exact matches? This would be way more intuitive.

**Comment by ceever** [comment]
@vshcherb: I concur with @scaidermern .. Why is there even a search button in the first place when I get all the results anyway whenever typing more than two letters? For the two letter case? A use case no user will ever understand?

I don't know—I would really suggest to overthink the whole instant search, search button and few letter case again. I feel many users will not understand how to work with the current setup.

I also suggest to review past issues here on Github regarding the search functionality again. Since my beginning with OSMand I had a bad feeling about the search functionality, which sometimes comes up with unexpected results. Just to mention the "Search nearby" functionality, which sometimes works far away and sometimes doesn't. It is just a little frustrating to have such erratic behaviour sometimes.

Nevertheless, thanks for the continues effort and investment 😇

**Comment by vshcherb** [comment]
As opposite to Online Services which could display results based on their relevancy (statistics what's been chosen most often) we can't do that unfortunately, so mixed approach has a systematic flaw in Offline app like OsmAnd.
![telegram-cloud-photo-size-4-5882253291977554392-y](https://user-images.githubusercontent.com/1042025/210346366-5df78d62-ee77-4f55-b462-e4b1cb3f6fd2.jpg)

**Comment by sonora** [comment]
The conventional name for that mechanism is "Search suggestions". You can try e..g. the web search of

- **Startpage.com**: They handle this like OsmAnd, and start flashing suggestions after 3 characters
- **google.com**: They have more bandwidth and compute power, and start flashing suggestions after the first character

In both cases:
- There is never an artificial delay!
- Typing a blank is the (natural) convention that you are looking for the complete expression, I guess we can keep this!

![2](https://user-images.githubusercontent.com/845735/210348072-098a6572-885a-4daf-9cf3-360e5d77ce21.jpg) ![1](https://user-images.githubusercontent.com/845735/210348067-7f007806-976d-4b5a-96cb-1df78e26f20f.jpg)
![3](https://user-images.githubusercontent.com/845735/210348074-48f720b5-5b50-4e5d-a6d1-72c7618e748c.jpg) 
![4](https://user-images.githubusercontent.com/845735/210348075-e2d7800b-5e04-44c4-9d53-f742cb150bcc.jpg)

**Comment by scaidermern** [comment]
> As opposite to Online Services which could display results based on their relevancy (statistics what's been chosen most often) we can't do that unfortunately

I'm not sure if this is already done but OsmAnd could priorize results that have a `wikipedia` or `wikidata` key. I think this is what Nominatim does (could be more complex, though). Similarly, other keys like `website` or `tourism` could also be used to increase the priority of these elements.

**Comment by scaidermern** [comment]
> The conventional name for that mechanism is "Search suggestions". You can try e..g. the web search of
> 
>     * **Startpage.com**: They handle this like OsmAnd, and start flashing suggestions after 3 characters

The main difference is that Startpage displays no suggestions if only two characters are present whereas OsmAnd already starts to list resultss
